rfcs/006-package-management.md

```diff
@@ -0,0 +1,243 @@
+conflict with something real.
+
+## Import statements
+
+The manifest file assigns a name to each dependency; to import the dependency
+named `foo` you simply write `import foo`. That is, an `import` statement either
+takes a string in quotes -- in which case it imports a path -- or an identifier
+without quote -- in which case it imports a package.
+
+The `import foo` expression evaluates to the contents of `main.ncl` in `foo`'s
+root directory.
+
+### Question: other entry points?
+
+We've hardcoded `main.ncl` as the entry point of every package, but what if
+they want to expose multiple entry points? For example, node allows a package's
+manifest to specify the entry point(s). This is probably not very important to
+support, as you can just put
+
+```nickel
+{
+  other = import "./other.ncl",
+  blah = import "./blah.ncl",
+}
+```
+
+in your package's `main.ncl`, to provide "other" and "blah" as other entry points.
+
+## Kinds of dependencies
+
+Where can dependencies come from? Dhall allows imports from arbitrary urls. Nix
+supports fetching from a variety of VCSs, paths, and archive formats.
+
+We'll support dependencies from
+
+- a central registry, that can identify packages by name and version number.
+  This should be the most common method of importing packages, like `crates.io`
+  in rust.
+- git repositories (either from HEAD, or from branches, tags, and revisions
+  specified by hashes). This allows for easy use of unpublished packages,
+  including in-development versions.
+- paths (relative or absolute). This allows for easy use of different packages
+  within the same repository, or for temporary patching of published packages.
+
+We require packages to have their own manifest file (at the package root), even
+if they don't import dependencies.
+
+## Lock files
+
+In order to ensure reproducibility across time and across machines, we build
+a lock-file (if there isn't yet one) when running `nickel eval` or `nickel
+export`. The lock-file specifies the exact versions of all (transitive)
+dependencies, allowing those identical versions to be used every time.
+
+- For a git dependency, the manifest might not specify the exact revision (it
+  might specify a branch or tag, or just default to HEAD). The lock-file will
+  record the exact revision.
+- For a repository dependency, the version specifier might allow for a range of
+  versions. The lock-file will record the exact version used.
+- For a path dependency, the lock-file will record that there was a path
+  dependency, but it won't record anything about it and it will ignore recursive
+  dependencies. This is because path dependencies can change at any time, so
+  they can't be meaningfully locked.
+
+What happens if we have a lock-file, but we modify the manifest? We don't want
+to be too strict about requiring the exact versions in the lock-file, or we'll
+end up forcing the user to re-create the lock-file from scratch. In this case,
```

Comment:
I'm not sure about this one. Not respecting the lockfile sounds like a hug avenue for reproducibility bugs (that is clearly stated as an overarching goal in the introduction: if it works on my machine, it works on your machine). Do you have example of such behavior in other mainstream language-specific package managers?

Reply:
Cargo has this behavior (which is where I first learned it from). For example, if I make a new crate with a dependency on `serde = "1"` then it will generate a lock-file with `serde` 1.0.203. If I then edit my `Cargo.toml` to specify `serde = "=1.0.202"` and run `cargo build`, it will automatically re-create the lock-file with `serde` version 1.0.202.

I think the reason this doesn't kill reproducibility too badly is that it only gets triggered if a manifest file changes. If I have a working lock-file and you copy my files to your machine, nickel's first attempt to resolve versions will try the same ones as in the lock-file. Those choices will all succeed, and so you'll get the exact same versions as me. The "lock-file-as-a-suggestion" path only gets triggered if there was a conflict when trying to use the lock-file's versions.

Unfortunately I don't know enough about other package managers' corner cases to know how widespread this behavior is. I think poetry doesn't do it, because it only modifies the lock-file when you explicitly ask it to. I'll investigate some others...

Reply:
Ok, so maybe a different way of saying it, is that if there is a conflict between the manifest and the lockfile, the lockfile is automatically re-generated. Which I think might be ok.

About cargo, I really don't like that `cargo install` from a published crate entirely ignores the lockfile for example (even if it agrees with the manifest), and that has been a bad user experience to try to understand why it's been failing with `install` but not with `build` locally. While cargo is great, we should make sure we aren't too biased and we don't just blindly copy their decisions just because we work in Rust these days (and constraints might be quite different for a Rust package and a Nickel package, although to be fair, the lockfile/manifest sync sounds like a very generic problem)

Reply:
> Ok, so maybe a different way of saying it, is that if there is a conflict between the manifest and the lockfile, the lockfile is automatically re-generated.

One subtlety here is that the lockfile isn't regenerated from scratch. It's regenerated by trying to run the previous resolution, but then backtracking when conflicts occur. Most of the time this should result in a fairly small change to the resulting resolution, as opposed to re-resolving from scratch which might bump lots of versions.